File: 110. Balanced Binary Tree.py
#Time Complexity:O(n)
#Space Complexity:O(h)
class Solution(object):
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if root is None:
            return True
        if(abs(self.dfs(root.left)-self.dfs(root.right))>1):
            return False
        return self.isBalanced(root.left) and self.isBalanced(root.right)
    
    def dfs(self,root):
        if root==None:
            return -1
        case1=self.dfs(root.left)
        case2=self.dfs(root.right)
        return max(case1,case2)+1

File: _110_BalancedBinaryTree.java
// Time Complexity :o(n) where n is number of nodes
// Space Complexity : o(n) where n is number of nodes (use of implicit stack)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : differnce in height of leftsubtree and right subtree should not be more than 1 


// Your code here along with comments explaining your approach :
class Solution {
    public boolean isBalanced(TreeNode root) {
      int height = getHeightOfTree(root);
      return height != -1;
    }

    public int getHeightOfTree(TreeNode root){

        if(root == null) return 0;

        int left = getHeightOfTree(root.left);
        int right = getHeightOfTree(root.right);

        if(left == -1 || right  == -1 || Math.abs(right - left) > 1) return -1;

        return Math.max(left, right) + 1;

    }
}
